Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrMetadataEmitter.kt

```diff
@@ -20,313 +21,460 @@ class StubIrMetadataEmitter(
+                    container = element,
+                    uniqIds = data.uniqIds.createChild(element.classifier.relativeFqName),
+                    typeParametersInterner = Interner(data.typeParametersInterner)
+            )
+            val elements = KmElements(element.children.map { it.accept(this, classVisitingContext) })
+            val kmClass = with (MappingExtensions(data.typeParametersInterner)) {
+                KmClass().also { km ->
+                    element.annotations.mapTo(km.annotations) { it.map() }
+                    km.flags = element.flags
+                    km.name = element.classifier.fqNameSerialized
+                    element.superClassInit?.let { km.supertypes += it.type.map() }
+                    element.interfaces.mapTo(km.supertypes) { it.map() }
+                    element.classes.mapTo(km.nestedClasses) { it.nestedName() }
+                    km.typeAliases += elements.typeAliases.toList()
+                    km.properties += elements.properties.toList()
+                    km.functions += elements.functions.toList()
+                    km.constructors += elements.constructors.toList()
+                    km.companionObject = element.companion?.nestedName()
+                    km.uniqId = data.uniqIds.uniqIdForClass(element)
+                }
             }
-            getterAnnotations += when (element.kind) {
-                is PropertyStub.Kind.Val -> element.kind.getter.annotations.map(AnnotationStub::map)
-                is PropertyStub.Kind.Var -> element.kind.getter.annotations.map(AnnotationStub::map)
-                is PropertyStub.Kind.Constant -> emptyList()
-            }
-            if (element.kind is PropertyStub.Kind.Constant) {
-                compileTimeValue = element.kind.constant.map()
-            }
-        }.let(properties::add)
-    }
-
-    override fun visitConstructor(constructorStub: ConstructorStub, data: StubContainer?) {
-        // TODO("not implemented")
-    }
+            // Metadata stores classes as flat list.
+            return listOf(kmClass) + elements.classes
+        }
 
-    override fun visitPropertyAccessor(propertyAccessor: PropertyAccessor, data: StubContainer?) {
-        // TODO("not implemented")
-    }
+        override fun visitTypealias(element: TypealiasStub, data: VisitingContext): KmTypeAlias =
+                with (MappingExtensions(data.typeParametersInterner)) {
+                    KmTypeAlias(element.flags, element.alias.topLevelName).also { km ->
+                        km.uniqId = data.uniqIds.uniqIdForTypeAlias(element)
+                        km.underlyingType = element.aliasee.map(shouldExpandTypeAliases = false)
+                        km.expandedType = element.aliasee.map()
+                    }
+                }
+
+        override fun visitFunction(element: FunctionStub, data: VisitingContext) =
+                with (MappingExtensions(data.typeParametersInterner)) {
```

Comment:
Does this approach handle the case when function type parameters are used in function signature?

Reply:
In general, yes. But current implementation is incorrect :)